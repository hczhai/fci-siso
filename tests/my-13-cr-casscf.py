import os
import sys

sys.path[:0] = [os.path.join(os.path.dirname(
    os.path.abspath(__file__)), "..")]

from pyscf import gto, scf, mcscf
from fcisiso import FCISISO
import numpy as np

mol = gto.M(atom='Cr 0 0 0',
            symmetry=False,
            basis='ccpvtz-dk', spin=1, charge=5, verbose=3)
mf = scf.newton(scf.RHF(mol).sfx2c1e())
mf.kernel()

print("\n Molecular orbital components:\n")

# total: 19 elec
# core: 1s2 2s2 2p6 3s2 3p6 (18e, 9o)
# active: 3d1 (1e, 5o)
ao_lb = gto.sph_labels(mol)
for i in range(40):
    ivs = np.argsort(mf.mo_coeff[:, i]**2)
    ss = ""
    for iv in ivs[::-1][:3]:
        ss += "%10.6f (%s)" % (mf.mo_coeff[iv, i] ** 2, ao_lb[iv].strip())
    print(" MO %2d :: %s" % (i, ss))

print("\n CASSCF(1, 5):\n")

cas_list = [10, 11, 12, 13, 14]
mc = mcscf.CASSCF(mf, 5, 1)
mc.fix_spin_(ss=0.75)
mo = mcscf.sort_mo(mc, mf.mo_coeff, cas_list, base=0)
mc.state_average_((0.2, 0.2, 0.2, 0.2, 0.2))
mc.kernel(mo)

print(mc.nelecas, mc.e_states)

dmao = mc.make_rdm1()
print(dmao.shape)

print("\n New active orbitals:\n")

for i in range(10, 30):
    ivs = np.argsort(mc.mo_coeff[:, i]**2)
    ss = ""
    for iv in ivs[::-1][:3]:
        ss += "%10.6f (%s)" % (mc.mo_coeff[iv, i] ** 2, ao_lb[iv].strip())
    print(" MO %2d :: %s" % (i, ss))

states = [(5, 2)]
ci = [
    (*mc.nelecas[::-1], 1, -1, mc.e_states[0], mc.ci[0].T),
    (*mc.nelecas[::-1], 1, -1, mc.e_states[1], mc.ci[1].T),
    (*mc.nelecas[::-1], 1, -1, mc.e_states[2], mc.ci[2].T),
    (*mc.nelecas[::-1], 1, -1, mc.e_states[3], mc.ci[3].T),
    (*mc.nelecas[::-1], 1, -1, mc.e_states[4], mc.ci[4].T),
    (*mc.nelecas, 1, 1, mc.e_states[0], mc.ci[0]),
    (*mc.nelecas, 1, 1, mc.e_states[1], mc.ci[1]),
    (*mc.nelecas, 1, 1, mc.e_states[2], mc.ci[2]),
    (*mc.nelecas, 1, 1, mc.e_states[3], mc.ci[3]),
    (*mc.nelecas, 1, 1, mc.e_states[4], mc.ci[4]),
]

siso = FCISISO(mol, mc, states)
siso.ci = ci
siso.ncore = 9
siso.norb = 5
energies = siso.kernel(dmao=dmao, amfi=True)

print(energies)

e0 = np.average(energies[0:4])
e1 = np.average(energies[4:10])

au2ev = 27.21139
print("")
print("E 2D(3/2)         = %10.4f eV" % (e0 * au2ev))
print("E 2D(5/2)         = %10.4f eV" % (e1 * au2ev))
print("2D(5/2) - 2D(3/2) = %10.4f eV" % ((e1 - e0) * au2ev))
print("2D(5/2) - 2D(3/2) = %10.4f cm-1" % ((e1 - e0) * au2ev * 8065.56))
