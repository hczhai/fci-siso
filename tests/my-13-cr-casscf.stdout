
WARN: HOMO -2.0095408750580166 > LUMO -2.7379072049185296 was found in the canonicalized orbitals.

converged SCF energy = -1042.42919675447

 Molecular orbital components:

 MO  0 ::   0.956710 (0 Cr 1s)  0.043281 (0 Cr 2s)  0.000000 (0 Cr 3s)
 MO  1 ::   0.959934 (0 Cr 2s)  0.043524 (0 Cr 1s)  0.000371 (0 Cr 4s)
 MO  2 ::   0.823188 (0 Cr 2px)  0.130237 (0 Cr 2pz)  0.046621 (0 Cr 2py)
 MO  3 ::   0.876285 (0 Cr 2py)  0.088490 (0 Cr 2px)  0.035270 (0 Cr 2pz)
 MO  4 ::   0.834539 (0 Cr 2pz)  0.088368 (0 Cr 2px)  0.077140 (0 Cr 2py)
 MO  5 ::   0.650076 (0 Cr 3s)  0.218378 (0 Cr 4s)  0.059279 (0 Cr 5s)
 MO  6 ::   2.290696 (0 Cr 4pz)  2.093662 (0 Cr 5pz)  0.901317 (0 Cr 3pz)
 MO  7 ::   2.341495 (0 Cr 4px)  2.140092 (0 Cr 5px)  0.921305 (0 Cr 3px)
 MO  8 ::   2.374846 (0 Cr 4py)  2.170574 (0 Cr 5py)  0.934428 (0 Cr 3py)
 MO  9 ::   3.203898 (0 Cr 5s)  0.485187 (0 Cr 4s)  0.184874 (0 Cr 3s)
 MO 10 ::   0.329360 (0 Cr 3dyz)  0.312104 (0 Cr 3dx2-y2)  0.160068 (0 Cr 3dxz)
 MO 11 ::   0.584766 (0 Cr 3dxz)  0.120097 (0 Cr 3dxy)  0.104968 (0 Cr 3dx2-y2)
 MO 12 ::   0.799878 (0 Cr 3dz^2)  0.066540 (0 Cr 4dz^2)  0.030165 (0 Cr 3dx2-y2)
 MO 13 ::   0.394963 (0 Cr 3dx2-y2)  0.265814 (0 Cr 3dyz)  0.166310 (0 Cr 3dxy)
 MO 14 ::   0.517407 (0 Cr 3dxy)  0.241733 (0 Cr 3dyz)  0.082898 (0 Cr 3dxz)
 MO 15 ::   0.742134 (0 Cr 5px)  0.540670 (0 Cr 5py)  0.108151 (0 Cr 5pz)
 MO 16 ::   0.779136 (0 Cr 5py)  0.324619 (0 Cr 5pz)  0.287199 (0 Cr 5px)
 MO 17 ::   0.958184 (0 Cr 5pz)  0.361621 (0 Cr 5px)  0.129717 (0 Cr 6pz)
 MO 18 ::   0.369610 (0 Cr 4dxy)  0.229467 (0 Cr 4dxz)  0.189400 (0 Cr 4dyz)
 MO 19 ::   0.749370 (0 Cr 4dz^2)  0.130302 (0 Cr 3dz^2)  0.056128 (0 Cr 4dx2-y2)
 MO 20 ::   0.288300 (0 Cr 4dxz)  0.215128 (0 Cr 4dx2-y2)  0.137988 (0 Cr 4dxy)
 MO 21 ::   0.527284 (0 Cr 4dyz)  0.363736 (0 Cr 4dxz)  0.091685 (0 Cr 3dyz)
 MO 22 ::   0.527481 (0 Cr 4dx2-y2)  0.350673 (0 Cr 4dxy)  0.091720 (0 Cr 3dx2-y2)
 MO 23 ::   5.679346 (0 Cr 5s)  4.350143 (0 Cr 4s)  0.980693 (0 Cr 7s)
 MO 24 ::  15.864830 (0 Cr 4pz) 13.647077 (0 Cr 5pz)  8.539841 (0 Cr 4px)
 MO 25 ::  11.698415 (0 Cr 4pz) 10.063087 (0 Cr 5pz)  8.546961 (0 Cr 4px)
 MO 26 ::  16.835947 (0 Cr 4py) 14.482441 (0 Cr 5py) 10.727297 (0 Cr 4px)
 MO 27 ::   0.859174 (0 Cr 6dxy)  0.583539 (0 Cr 4dxy)  0.345693 (0 Cr 6dyz)
 MO 28 ::   1.416047 (0 Cr 6dz^2)  0.961760 (0 Cr 4dz^2)  0.322975 (0 Cr 3dz^2)
 MO 29 ::   0.922307 (0 Cr 6dyz)  0.626419 (0 Cr 4dyz)  0.370402 (0 Cr 6dxy)
 MO 30 ::   1.008202 (0 Cr 6dx2-y2)  0.684758 (0 Cr 4dx2-y2)  0.229953 (0 Cr 3dx2-y2)
 MO 31 ::   1.105839 (0 Cr 6dxz)  0.751071 (0 Cr 4dxz)  0.252222 (0 Cr 3dxz)
 MO 32 ::  45.889240 (0 Cr 4px) 31.320291 (0 Cr 5px)  6.331207 (0 Cr 4pz)
 MO 33 ::  23.308357 (0 Cr 4py) 22.585588 (0 Cr 4pz) 15.908403 (0 Cr 5py)
 MO 34 ::  29.286682 (0 Cr 4pz) 28.912090 (0 Cr 4py) 19.988725 (0 Cr 5pz)
 MO 35 ::  30.928285 (0 Cr 4s) 13.833725 (0 Cr 5s)  5.649961 (0 Cr 3s)
 MO 36 ::   0.649728 (0 Cr 5f+0)  0.151905 (0 Cr 5f-2)  0.112296 (0 Cr 5f+2)
 MO 37 ::   0.531809 (0 Cr 5f-1)  0.222974 (0 Cr 5f+1)  0.139056 (0 Cr 5f+3)
 MO 38 ::   0.440945 (0 Cr 5f+2)  0.149829 (0 Cr 5f+3)  0.116910 (0 Cr 5f-3)
 MO 39 ::   0.524040 (0 Cr 5f+3)  0.212886 (0 Cr 5f+2)  0.076615 (0 Cr 5f-1)

 CASSCF(1, 5):

CASSCF energy = -1043.49897057923
CASCI E = -1043.49897057923  E(CI) = -3.33411253825057  S^2 = 0.7500000
CASCI state-averaged energy = -1043.49897057923
CASCI energy for each state
  State 0 weight 0.2  E = -1043.49897057928 S^2 = 0.7500000
  State 1 weight 0.2  E = -1043.49897057927 S^2 = 0.7500000
  State 2 weight 0.2  E = -1043.49897057922 S^2 = 0.7500000
  State 3 weight 0.2  E = -1043.49897057921 S^2 = 0.7500000
  State 4 weight 0.2  E = -1043.49897057919 S^2 = 0.7500000
(1, 0) [-1043.49897058 -1043.49897058 -1043.49897058 -1043.49897058
 -1043.49897058]
(68, 68)

 New active orbitals:

 MO 10 ::   0.573229 (0 Cr 3dxz)  0.117728 (0 Cr 3dxy)  0.102897 (0 Cr 3dx2-y2)
 MO 11 ::   0.784097 (0 Cr 3dz^2)  0.077602 (0 Cr 4dz^2)  0.029570 (0 Cr 3dx2-y2)
 MO 12 ::   0.387171 (0 Cr 3dx2-y2)  0.260570 (0 Cr 3dyz)  0.163029 (0 Cr 3dxy)
 MO 13 ::   0.507199 (0 Cr 3dxy)  0.236963 (0 Cr 3dyz)  0.081262 (0 Cr 3dxz)
 MO 14 ::   2.452845 (0 Cr 5s)  0.232575 (0 Cr 4s)  0.141578 (0 Cr 3s)
 MO 15 ::   3.895439 (0 Cr 5px)  1.506909 (0 Cr 4px)  1.276339 (0 Cr 5pz)
 MO 16 ::   3.343291 (0 Cr 5py)  2.104824 (0 Cr 5pz)  1.293317 (0 Cr 4py)
 MO 17 ::   2.144835 (0 Cr 5pz)  1.828486 (0 Cr 5py)  1.552677 (0 Cr 5px)
 MO 18 ::   0.301189 (0 Cr 4dxy)  0.205641 (0 Cr 4dyz)  0.123865 (0 Cr 4dxz)
 MO 19 ::   0.359875 (0 Cr 4dxz)  0.241357 (0 Cr 4dxy)  0.130305 (0 Cr 4dyz)
 MO 20 ::   0.684156 (0 Cr 4dz^2)  0.132735 (0 Cr 3dz^2)  0.095070 (0 Cr 6dz^2)
 MO 21 ::   0.456165 (0 Cr 4dyz)  0.298902 (0 Cr 4dxz)  0.088502 (0 Cr 3dyz)
 MO 22 ::   0.570467 (0 Cr 4dx2-y2)  0.190429 (0 Cr 4dxy)  0.110678 (0 Cr 3dx2-y2)
 MO 23 ::   5.511448 (0 Cr 5s)  3.235015 (0 Cr 4s)  1.547516 (0 Cr 7s)
 MO 24 ::  36.771339 (0 Cr 4pz) 33.959027 (0 Cr 5pz)  7.382257 (0 Cr 4px)
 MO 25 ::  32.567871 (0 Cr 4px) 30.077045 (0 Cr 5px)  8.169762 (0 Cr 4pz)
 MO 26 ::  39.944732 (0 Cr 4py) 36.889716 (0 Cr 5py)  4.996369 (0 Cr 4px)
 MO 27 ::   0.870518 (0 Cr 6dxy)  0.741079 (0 Cr 4dxy)  0.267411 (0 Cr 6dyz)
 MO 28 ::   0.795667 (0 Cr 6dyz)  0.677358 (0 Cr 4dyz)  0.359290 (0 Cr 6dz^2)
 MO 29 ::   1.046425 (0 Cr 6dz^2)  0.890831 (0 Cr 4dz^2)  0.301937 (0 Cr 3dz^2)

Generating Spin-Orbit Integrals:

0.009432658358674316

Complex SO-Hamiltonian matrix elements over spin components of spin-free eigenstates:
(In cm-1. Print threshold:    29.000 cm-1)

  I1 =    1 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =  -213.842
  I1 =    2 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =  -120.981
  I1 =    2 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    1 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =  -139.613
  I1 =    3 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =   118.534
  I1 =    3 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    1 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =   -85.266
  I1 =    3 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =  -104.820
  I1 =    4 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =  -254.310
  I1 =    4 (E1 =  -1043.49897058) S1 =  0.5 MS1 = -0.5 I2 =    3 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     0.000 Im =   -34.455
  I1 =    5 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    1 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   248.249 Im =  -184.559
  I1 =    5 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -144.139 Im =   -84.684
  I1 =    5 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    3 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =    19.636 Im =  -124.074
  I1 =    6 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -248.249 Im =   184.559
  I1 =    6 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =    -4.563 Im =   142.251
  I1 =    6 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    3 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -144.829 Im =   -63.632
  I1 =    6 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    5 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =   213.842
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   144.139 Im =    84.684
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    1 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =     4.563 Im =  -142.251
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    3 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   129.568 Im =   -88.253
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    4 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -159.203 Im =    79.861
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    5 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =   120.981
  I1 =    7 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    6 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =   139.613
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    0 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   -19.636 Im =   124.074
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    1 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   144.829 Im =    63.632
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -129.568 Im =    88.253
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    4 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   179.835 Im =   288.229
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    5 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =  -118.534
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    6 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =    85.266
  I1 =    8 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    7 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =   104.820
  I1 =    9 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    2 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =   159.203 Im =   -79.861
  I1 =    9 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    3 (E2 =  -1043.49897058) S2 =  0.5 MS2 = -0.5 Re =  -179.835 Im =  -288.229
  I1 =    9 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    7 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =   254.310
  I1 =    9 (E1 =  -1043.49897058) S1 =  0.5 MS1 =  0.5 I2 =    8 (E2 =  -1043.49897058) S2 =  0.5 MS2 =  0.5 Re =     0.000 Im =    34.455

Total energies including SO-coupling:

  State    0 Total energy:  -1043.50155382 | largest |coeff|**2   0.398772 from I =    0 E =  -1043.49897058 S =  0.5
  State    1 Total energy:  -1043.50155382 | largest |coeff|**2   0.398800 from I =    0 E =  -1043.49897058 S =  0.5
  State    2 Total energy:  -1043.50155382 | largest |coeff|**2   0.399919 from I =    4 E =  -1043.49897058 S =  0.5
  State    3 Total energy:  -1043.50155382 | largest |coeff|**2   0.399910 from I =    4 E =  -1043.49897058 S =  0.5
  State    4 Total energy:  -1043.49724842 | largest |coeff|**2   0.509699 from I =    0 E =  -1043.49897058 S =  0.5
  State    5 Total energy:  -1043.49724842 | largest |coeff|**2   0.509831 from I =    0 E =  -1043.49897058 S =  0.5
  State    6 Total energy:  -1043.49724842 | largest |coeff|**2   0.489477 from I =    2 E =  -1043.49897058 S =  0.5
  State    7 Total energy:  -1043.49724842 | largest |coeff|**2   0.490112 from I =    2 E =  -1043.49897058 S =  0.5
  State    8 Total energy:  -1043.49724842 | largest |coeff|**2   0.587704 from I =    4 E =  -1043.49897058 S =  0.5
  State    9 Total energy:  -1043.49724842 | largest |coeff|**2   0.587018 from I =    4 E =  -1043.49897058 S =  0.5
[-1043.50155382 -1043.50155382 -1043.50155382 -1043.50155382
 -1043.49724842 -1043.49724842 -1043.49724842 -1043.49724842
 -1043.49724842 -1043.49724842]

E 2D(3/2)         = -28395.1277 eV
E 2D(5/2)         = -28395.0106 eV
2D(5/2) - 2D(3/2) =     0.1172 eV
2D(5/2) - 2D(3/2) =   944.9280 cm-1
